app/src/main/java/org/mozilla/fenix/utils/WaitableAsyncInflater.kt

```diff
@@ -0,0 +1,65 @@
+package org.mozilla.fenix.utils
+
+import android.content.Context
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.asynclayoutinflater.view.AsyncLayoutInflater
+
+/**
+ *
+ */
+class WaitableAsyncInflater(val resId: Int) {
+
+    var asyncInflater: AsyncLayoutInflater? = null
+    var condition = Object()
+    var view: View? = null
+
+    /**
+     *
+     */
+    private fun startAsynchronousInflation(context: Context, group: ViewGroup) {
+        val newInflater = AsyncLayoutInflater(context)
+        newInflater.inflate(
+            resId,
+            group,
+            AsyncLayoutInflater.OnInflateFinishedListener { newView: View,
+                                                            _: Int,
+                                                            _: ViewGroup? ->
+                view = newView
+                synchronized(condition) {
+                    condition.notifyAll()
+                }
+            })
+        asyncInflater = newInflater
+    }
+
+    private fun doSynchronousInflation(context: Context, group: ViewGroup) {
```

Comment:
I'm not sure you need to the `doSynchronousInflation` since `AsyncInflater` will default to do it on the MT if it can't do it async.

Ref: 

> [For a layout to be inflated asynchronously it needs to have a parent whose ViewGroup.generateLayoutParams(AttributeSet) is thread-safe and all the Views being constructed as part of inflation must not create any Handlers or otherwise call Looper.myLooper(). If the layout that is trying to be inflated cannot be constructed asynchronously for whatever reason, AsyncLayoutInflater will automatically fall back to inflating on the UI thread. ](https://developer.android.com/reference/androidx/asynclayoutinflater/view/AsyncLayoutInflater)

Reply:
Unless I misunderstood the use of the `doSynchronousInflation` 😛

Reply:
Thanks for the review! I _think_ we are talking about two different things. `WaitiableAsyncInflater` does this synchronously because it needs to be done immediately not because it needs to be done on the UI thread. Please tell me where I am wrong, but the way I read the document you quoted (which was really helpful to read, thank you!), `AsyncLayoutInflater` will properly handle whether or not it can do the actual inflation in a background thread or only on the UI thread but it will do the inflation asynchronously no matter what. 

`WaitiableAsyncInflater` has to make the distinction between doing it synchronously and asynchronously. In the case where `get()` invokes `doSynchronousInflation()`, the caller of `get()` needs the view but there is a) no view already inflated and b) no ongoing asynchronous action to create that view. Because of the way that the `AsyncLayoutInflater` notifies the callback that inflation is complete, `WaitiableAsyncInflater` cannot simply launch an `AsyncLayoutInflater` job in this case and wait for the callback because it will deadlock. 

This is the case: 
1. From the UI thread, a user of the `WaitiableAsyncInflater` calls `get()`.
2. There is no view and no ongoing asynchronous action to create it, so `WaitiableAsyncInflater` starts an asynchronous action and then blocks the UI thread until it is ready.
3. When `AsyncLayoutInflater` completes the asynchronous inflation, it posts a task to the UI thread to invoke the associated callback. 

Because `WaitiableAsyncInflater` is blocking the UI thread on the completion of the asynchronous inflation, the UI thread will not continue processing incoming tasks posted to it and, therefore, will never execute the callback. Boom.

Please, please scrutinize my logic to make sure that I am not saying anything you disagree with. Thanks again for a) the review, b) including the documentation and c) checking my logic.